xarray/tests/test_dataarray.py

```diff
@@ -1392,6 +1402,45 @@ def test_expand_dims_with_scalar_coordinate(self):
         roundtripped = actual.squeeze(['z'], drop=False)
         assert_identical(array, roundtripped)
 
+    def test_expand_dims_with_greater_dim_size(self):
+        """Python 3.6+ dicts keep insertion order, unlike Python 3.5 and
+        earlier. Therefore the following tests have to use ordered dicts to
+        pass for python 3.5 and earlier.
+        """
+        array = DataArray(np.random.randn(3, 4), dims=['x', 'dim_0'],
+                          coords={'x': np.linspace(0.0, 1.0, 3), 'z': 1.0},
+                          attrs={'key': 'entry'})
+        # For python 3.5 and earlier this has to be an ordered dict, to
+        # maintain insertion order.
+        actual = array.expand_dims(
+            OrderedDict((('y', 2), ('z', 1), ('dim_1', ['a', 'b', 'c']))))
+
+        expected_coords = OrderedDict((
+            ('y', [0, 1]), ('z', [1.0]), ('dim_1', ['a', 'b', 'c']),
+            ('x', np.linspace(0, 1, 3)), ('dim_0', range(4))))
+        expected = DataArray(array.values * np.ones([2, 1, 3, 3, 4]),
+                             coords=expected_coords,
+                             dims=list(expected_coords.keys()),
+                             attrs={'key': 'entry'}
+                             ).drop(['y', 'dim_0'])
+        assert_identical(expected, actual)
+
+        # Test with kwargs instead of passing dict to dim arg.
+        other_way = array.expand_dims(y=2, z=1, dim_1=['a', 'b', 'c'])
+        # Unfortunately, there is no way to maintain insertion order with
+        # kwargs in python 3.5 and earlier, so for now we have to ensure the
```

Comment:
Could we make specifying sizes for multiple dimensions with a non-ordered dictionary an error on Python 3.5? I don't want users to rely on behavior that will break in subtle ways when they rerun their code in a new Python session -- that is a very poor user experience.

Reply:
That's a good point and a good solution, thanks!

Reply:
Would if I always sort `dim`/`dim_kwargs` to ensure the order will be the same between python 3.5 and python 3.6. This would of course be clearly documented in the doc string.

So I think it would just be
```
        if isinstance(dim, int):
            raise TypeError('dim should be str or sequence of strs or dict')
        elif isinstance(dim, str):
            dim = {dim: 1}
        elif isinstance(dim, (list, tuple)):
            if len(dim) != len(set(dim)):
                raise ValueError('dims should not contain duplicate values.')
            dim = {d: 1 for d in dim}
        dim = either_dict_or_kwargs(dim, dim_kwargs, 'expand_dims')

        dim = sorted(dim)
```

Reply:
I guess the problem with the above is that will change the output for current users.

Reply:
I would rather not sort. It's more intuitive for users if dimensions appear in the provided order.

At this point Python 3.5 is the oldest Python version we support (there is also 3.6 and 3.7). I don't think we should go out of our way to make things more convenient for Python 3.5 users at the cost of inconveniencing the majority of our users who use newer Python versions. See https://pypistats.org/packages/xarray if you're curious about download numbers.